Improve subject hierarchy combobox

agirault opened the pull request:
- [x] Option for new vertical alignment of the popup treeview
- [x] Disable selections of parent items when using attribute filter
- [x] Do not hide popup when no item was selected
- [x] Implement qMRMLSubjectHierarchyComboBox::clearCurrentItem 
- [x] Update combobox label on SubjectHierarchyItemModifiedEvent

In a future pr:
- [ ] Scene item shown when root item != scene and filters returns no item
- [ ] combobox does not stretch in layout like qMRMLNodeComboBox
- [ ] popup wider that combobox due to margins

agirault commented:
Related to [Disable selections of parent items when using attribute filter](52419ff):
1. In [`onSelectionChanged`](https://github.com/Slicer/Slicer/pull/813/commits/52419ffd9082a037a1381eafb1ef9bc5d737f61d#diff-784aa1951ed8cd51bfe8a1bb9fe2ee5bR848) of `qMRMLSubjectHierarchyTreeView`, I catch the items that should not have been selected (to do this, I had to make [`filterAcceptsItem`](https://github.com/Slicer/Slicer/pull/813/commits/52419ffd9082a037a1381eafb1ef9bc5d737f61d#diff-43b6ab175e21e11004c1b5f3760efd7dR97) in `qMRMLSortFilterSubjectHierarchyProxyModel` public, and add a parameter `acceptIfAnyChildIsAccepted`, defaulting to true to keep the previous behavior, but set to false in this case). Then, if nothing is selected, I select the scene item if nothing was blocked (previous default behavior), but if something was blocked, I think we should reselect the deselected items: for example, if I had a study selected, and then select the patient, the popover menu should not close and the study should be re-selected. Right now I do not yet know how to do this using the `QItemSelection deselected`).

2. In [`qMRMLSubjectHierarchyTreeView::mousePressEvent`](https://github.com/Slicer/Slicer/pull/813/commits/52419ffd9082a037a1381eafb1ef9bc5d737f61d#diff-784aa1951ed8cd51bfe8a1bb9fe2ee5bR779) I only forward the event to the base class `QTreeView` if `filterAcceptsItem` (with `acceptIfAnyChildIsAccepted=false`). This avoid the item from even getting selected when pressed, but due to this, the branch control ([-] or [+] on the left of the item to contract/expand the branch) is not working anymore. Could investigate the `Qt::ItemIsSelectable` flag instead?

agirault commented:
Related to [Disable selections of parent items when using attribute filter](52419ff):

In `onSelectionChanged` of [`qMRMLSubjectHierarchyTreeView`]((https://github.com/Slicer/Slicer/pull/813/commits/02df4a6baf86b7c6c0bbc6627df34b2696093fbb#diff-784aa1951ed8cd51bfe8a1bb9fe2ee5b)), I catch the items that should not have been selected (to do this, I had to make `filterAcceptsItem` in [`qMRMLSortFilterSubjectHierarchyProxyModel`](https://github.com/Slicer/Slicer/pull/813/files#diff-43b6ab175e21e11004c1b5f3760efd7d) public, and add a parameter `acceptIfAnyChildIsAccepted`, defaulting to true to keep the previous behavior, but set to false in this case). Then, if nothing is selected, I select the scene item if nothing was blocked (previous default behavior), but if something was blocked, I think we should reselect the deselected items: for example, if I had a study selected, and then select the patient, the popover menu should not close and the study should be re-selected. Right now I do not yet know how to do this using the `QItemSelection deselected`).

I thought about only forwarding the event to the base class `QTreeView` in [`qMRMLSubjectHierarchyTreeView::mousePressEvent`](https://github.com/Slicer/Slicer/blob/cb1c7208464d13fb337659dadc62af7ee6fc25f4/Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyTreeView.cxx#L776) if `filterAcceptsItem` (with `acceptIfAnyChildIsAccepted=false`), to avoid the item from even getting selected when pressed, but due to this, the branch control ([-] or [+] on the left of the item to contract/expand the branch) would not be working anymore. Could investigate the `Qt::ItemIsSelectable` flag instead?

cpinter commented:
After a chat with @agirault we have a better plan for the "Disable selections of parent items when using attribute filter" part. The setXFilter methods would call a function called updateItemSelectableFlags that goes through each item and sets the selectable flag to false if the filter rejects it (ignoring the parents, as it is already in this PR). To make sure that the flags are correct after a full update, the model needs to be connected to the tree view (in TreeViewPrivate::init) like this
QObject::connect( this->Model, SIGNAL(subjectHierarchyUpdated()), q, SLOT(updateItemSelectableFlags ()) );

This way the user won't have the chance to select items not accepted by the filter in the first place.

agirault commented on Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyTreeView.cxx:
```diff
@@ -603,6 +606,37 @@ void qMRMLSubjectHierarchyTreeView::updateRootItem()
+  foreach(vtkIdType itemID, childItemIDs)
+    {
+    bool isSelectable = this->sortFilterProxyModel()->filterAcceptsItem(itemID, false);
+    QStandardItem* item = sceneModel->itemFromSubjectHierarchyItem(itemID); // or proxy model?
```
- it seems that the `QStandardItems` are shared between treeviews,
so changing their selection impacts all the other treeviews
- is it ok to use `itemFromSubjectHierarchyItem` from the scenemodel?
Should there be a mapping with the sortFilterProxyModel? The latter
however does not have a `itemFromSubjectHierarchyItem` method.

cpinter replied:
* Cannot be possible. QStandardItems are the view items that are displayed in each treeview separately. Please read http://doc.qt.io/qt-4.8/model-view-programming.html and http://doc.qt.io/qt-4.8/modelview.html
* Using that function from the scene model when there is a proxy filter will probably make the application crash. Use the proxy model's converters instead.

cpinter replied:
What is strange that the tree views create their own model and proxymodel in qMRMLSubjectHierarchyTreeViewPrivate::init. I don't see how it is possible that the filters are applied to all of the views. Also I cannot reproduce this behavior, the filters are correctly applied on the separate widgets individually

agirault replied:
> Use the proxy model's converters instead.

What do you mean by converter here? Are you saying I should implement `itemFromSubjectHierarchyItem` in the proxy model class, and properly use the mapping within it?

cpinter replied:
If it's not there yet then yes please. You only need to call the same mapping function as the others and call the scene model's converter.

dzenanz commented:
@agirault This does not seem to be working correctly. We can talk on Monday if you need me to show what is broken.

cpinter commented:
@dzenanz This is WIP. Please wait a bit until @agirault implements what we have discussed and I explained above.

dzenanz commented:
He requested a review, that's why I tested this - even though it looked like it was not finished. I could not contribute constructively towards code changes, so I tried compiling and running it.

agirault commented:
@dzenanz, thanks. I do not expect it to work yet, the commit and pr are both marked WIP. I added you as a reviewer along with Csaba so you could give your inputs if you had any, given that this was for your project.

I'll work with Csaba's help to put this together properly, and I'll keep you posted then. 

Thank you @cpinter

agirault commented on Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSortFilterSubjectHierarchyProxyModel.cxx:
```diff
@@ -198,6 +198,14 @@ QModelIndex qMRMLSortFilterSubjectHierarchyProxyModel::indexFromSubjectHierarchy
+{
+  qMRMLSubjectHierarchyModel* sceneModel = qobject_cast<qMRMLSubjectHierarchyModel*>(this->sourceModel());
+  QModelIndex index = this->mapFromSource(sceneModel->indexFromSubjectHierarchyItem(itemId, column));
+  return sceneModel->itemFromIndex(index);
```
@cpinter is that what you were envisioning? This currently returns a null pointer in my use case for every item [here](https://github.com/Slicer/Slicer/pull/813/files#diff-784aa1951ed8cd51bfe8a1bb9fe2ee5bR628), so i'll keep investigating.

cpinter replied:
First of all, please don't reimplement anything that is there and works, i.e.
https://github.com/Slicer/Slicer/blob/master/Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSortFilterSubjectHierarchyProxyModel.cxx#L195

About the actual issue, I'm not sure if using the scene model's itemFromIndex function on an index coming form the proxy model would work. I'll take a deeper look.

cpinter replied:
This call works for the highlights
https://github.com/Slicer/Slicer/blob/master/Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyTreeView.cxx#L1247
I wonder why it didn't work for you. Apparently the scene model can be used to get the items of the columns from SH ID. I'd start from here. Thanks!

agirault commented:
The issue was with the way I was setting the flag: I needed `|` when adding `selectable`, and `&` to remove it ([here](https://github.com/Slicer/Slicer/pull/813/files#diff-784aa1951ed8cd51bfe8a1bb9fe2ee5bR635)). The initial way I retrieved the model index and item was correct, there was indeed no need for any mapping.

Still putting WIP since the combobox closes when selecting an unselectable item (but does not show it as the current item, so that's a good sign, almost there).

cpinter commented:
Sounds great, thanks for the update!
Yes, getting the item is apparently fine from the scene model, but the index needs to be retrieved through the proxy model. Sorry about the confusion, it was a long time since I implemented these, and the inner feelings of Qt are quite complex when it comes to view/models!

agirault commented:
No problem at all, thanks for the directions. Learning some of that too is good for me.

Following up on the combobox closing but empty: even if the selectable flag is set to false, clicking on an unselectable item will trigger the `selectionChanged`, since it seems to deselect the previously selected items (but the list of selected items is empty). Right now, this will result in the scene item being selected in `onSelectionChanged`

agirault commented:
@cpinter I found a cleaner and more robust way to do this by subclassing [`QAbstractItemModel::flags`](http://doc.qt.io/qt-4.8/qabstractitemmodel.html#flags) which ensure all items in a same row have the proper flags.

As adviced in Qt doc here: http://doc.qt.io/qt-4.8/qabstractitemmodel.html#subclassing
> To enable editing in your model, you must also implement setData(), and reimplement flags() to ensure that `ItemIsEditable` is returned. 

Two things remain up to discussion:
1. Do you think there is any reason to call `Superclass::flags` in [here](https://github.com/Slicer/Slicer/pull/813/files) ?
2. Because of [this line](https://github.com/Slicer/Slicer/blob/master/Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyTreeView.cxx#L867), the scene is set as the selected item when nothing is selected. Returning directly instead would make my code work (combobox stays open, unselectable row does not get selected), but that could be an issue when we try to unselect an item in a treeview, wouldn't it?

cpinter commented:
Selected and deselected are unused because it was easier to get the selection and set it directly instead of doing it "incrementally" as selection happens. It could be used as well, but it was not how I implemented the function.

OK, if subclassing works, then I don't have any objections going that way.

> 1. Do you think there is any reason to call Superclass::flags in here ?
Where? The link points to the whole diff

> 2.
Actually I don't know why the scene item ID needs to be the selected one instead of the invalid item ID. In any case, `emit currentItemChanged(newCurrentItemID);` needs to be called, and the highlight needs to be refreshed (cleared). You may try changing it to invalid item ID, but then it needs to be tested extensively. Do you know exactly which line makes the popup close?

agirault commented:
Well, `emit currentItemChanged(newCurrentItemID);` is what makes the popup close, because it acknowledged a change of selection, after updating it with `
  qSlicerSubjectHierarchyPluginHandler::instance()->setCurrentItems(selectedShItems);` that's why I wanted to return before selecting the scene as the current item and emitting that signal.

cpinter commented:
It the selection has not changed, that is nothing has been selected or deselected, then it does not need to be emitted. So if an unselectable item has been clicked, then as far as I understand it's not needed to be hit.

agirault commented:
> It the selection has not changed, that is nothing has been selected or deselected, then it does not need to be emitted. So if an unselectable item has been clicked, then as far as I understand it's not needed to be hit.

Agreed, what I'm afraid of is that `onSelectionChanged` is still called when I click on a non-selectable item, and I'm not sure why, or how to check if a selection has changed intentionally (clearSelection), or not (clicking on an unselectable item).

I'll investigate more.

cpinter commented:
You said that if an unselectable item was clicked, then the `selected` and `deselected` containers are both empty. Maybe that can be used for deciding this.

agirault commented:
That was incorrect: `deselected` will have the previous id of the item that was previously selected. 

To summarize, with my change, `selected` is now empty (flag helped not selecting the parent item), but `deselected` is the same (clicking on the parent item will still deselect the previous item). What I could do would be to reselect the item from `deselected`, but then I assume that would break any calls to `clearSelection`.

The best thing I can think of would be to avoid unselecting that initial item when clicking on the parent item, to address the problem at its source, I'm just not sure yet how to do this but I'll look more.

cpinter commented:
Sounds reasonable, although it's hard to imagine what it will be like when actually using it.

Maybe we're (almost) back where we started: re-adding the selection enclosed in this->blockSignals(true/false); statements when only deselection happens and nothing else is selected.
Please test the regular tree view as well (in Data module) - with single and multi-selection too - to make sure nothing unexpected happens when clicking around. Thanks!

agirault commented:
>  when only deselection happens and nothing else is selected.

My issue is that I don't think this is a strong enough condition since that it will be valid too when calling `clearSelection`, which happens when programmatically setting items:
- https://github.com/Slicer/Slicer/blob/master/Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyTreeView.cxx#L418
- https://github.com/Slicer/Slicer/blob/master/Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyTreeView.cxx#L446

cpinter commented:
When that condition is fulfilled, you can clear the cache container (SelectedItems) to be safe, but when the selection is made in setCurrentItem, then onSelectionChanged will run as expected and populate the container again.

agirault commented:
Here is to illustrate the issue I am trying to bring across:

![1](https://user-images.githubusercontent.com/4910855/31639176-608f96b0-b2a5-11e7-8a70-5784c4f017bc.gif)

What you can see here, is that - with the change we discussed - the shnodes do not get unselected when clicking outside of an item, since it reselects the deselected items if there are no selected items at the end of the selectionChanged.

That's bad if we want to reset a combobox for example, so to clear any current item (to show again the "Select subject hierarchy item")

lassoan commented:
I think in a list or tree widget you cannot remove selection in a list by clicking in the empty area of a widget. Have you seen any widget where you could remove item selection like that?

jcfr commented:
I  do not have the full picture and did not thoroughly read the entire thread , but to programmatically unselect items,  I call `clearSelection`. See [here](https://github.com/commontk/CTK/blob/89071902e2746fca879e4ed8bde51567261eeeae/Libs/Widgets/ctkFileDialog.cpp#L218) for an example.

Depending on the selection mode, it is possible to have no item selected. For example, using the CTRL key it is possible to unselect all tiems. See [QAbstractItemView::SelectionMode](http://doc.qt.io/qt-4.8/qabstractitemview.html#SelectionMode-enum) and [QItemSelectionModel::Clear](http://doc.qt.io/qt-4.8/qitemselectionmodel.html#SelectionFlag-enum).

lassoan commented:
Yes, unselect with Control key is the method usually available in widgets. Clicking in empty area usually just changes the focus.

cpinter commented:
@agirault A combobox always have to ahve a selected item. This is why the MRML node comboboxes have a None option.

agirault commented:
@jcfr @lassoan thanks a lot for your input.

In the current master, Tree view DOES unselect its selected items when pressing something else than an item:
![2](https://user-images.githubusercontent.com/4910855/31667789-69bcec28-b31e-11e7-9838-8a0a427bde9a.gif)

To be specific, what it does is catching a `deselect` in `onSelectionChanged` [here](https://github.com/Slicer/Slicer/blob/master/Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyTreeView.cxx#L837), then selects the scene item id to ensure something is selected, [here](https://github.com/Slicer/Slicer/blob/master/Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyTreeView.cxx#L867-L874), then emits a current item changed signal for INVALID_ITEM_ID [here](https://github.com/Slicer/Slicer/blob/master/Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyTreeView.cxx#L885-L890).

In this PR, it used to do the same when tapping on a parent item (which was disabled for selection because it doesn't pass the filters). You can see on the following gif that studies are selectable and that selecting patients (disabled for selection) will select the scene and emit a change, therefore closing the treeview popup:
![3](https://user-images.githubusercontent.com/4910855/31667992-f4660eb8-b31e-11e7-88b4-6b128da1b652.gif)

It defeats the purpose of maintaining the previous selection when clicking on something that we want to disable for selection. This is what I'd want to achieve:

![4](https://user-images.githubusercontent.com/4910855/31677280-455ad5a6-b338-11e7-8e08-530a1cdea4d7.gif)

However, doing this implies reselecting all deselected items when no selected items are shown, which results in the first gif I shared of the data module, where items can never be deselected:

![1](https://user-images.githubusercontent.com/4910855/31668433-332ee7f4-b320-11e7-93c4-9eb828fcbb61.gif)

... which also means that we would not be able to reset a subject hierarchy combobox to the message "Select subject hierarchy item" (could do so by setting an invalid item id), which I need to do in my own use case: when I change a study in the top combobox (`Study`), I want the comboboxes below (`Image1` and `Image21) to set their root item to that study, and to clear their previous selection, which can not happen anymore. You can see that in the third gif of this comment, `Image2` combobox does not reset to "Select subject hierarchy item" when I select a different study in the `Study` combobox.

cpinter commented:
> which also means that we would not be able to reset a subject hierarchy combobox to the message "Select subject hierarchy item"

This is why I said above that comboboxes do not support "none" selection. In Slicer when it is desirable it is achieved by adding a None item.
In your case you want a selection to determine another combobox's root item. I assume you set the root item in a slot that handles the currentItemChanged signal of the study combobox. If the new study does not contain the previously selected volume of the Image combobox, then the text it shows does not change? If that's the case I think it isn't that the selection didn't change, but simply that the text shown by the combobox didn't change, which is then a bug in this function
https://github.com/Slicer/Slicer/blob/master/Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx#L340

Furthermore, if handling the selection should be different in the tree view and the combobox, then the combobox could implement its own onSelectionChanged function, and connect it to the tree view's signal instead of the tree view's slot.

agirault commented:
I updated, cleaned up and squashes the commits. Now selecting a disabled item will deselect the previously selected item (like default), but the popup containing the treeview will not hide, to encourage the user to select a selectable item.

This allows to clear a combobox (which will set the scene as the current item, and emit an item changed signal to an invalid id to reset the combobox).

Also added a fix to ensure that the combobox text would get updated when a subject hierarchy item would get modified.

I could still work on one thing: it seems that a treeview with a root item different than the scene (using `setRootItem`) and with filters filtering out all items (using `setAttributeFilter`) will not display the proper items (which should be no items, or the root item not selectable if `showRootItem`). Instead, it displays the `Scene` item, and the app will crash when selecting it.

cpinter commented:
@agirault Thank you so much for working on this! I'll build your branch and try it out.

The root item issue won't be easy to fix I think; I struggled with the root item a lot already. If this is something that is present in the master, then I suggest adding a Mantis issue about it for now, and you and/or I can look at it after this PR is merged.

agirault commented:
We're going to need that fixed for our work,since it makes Slicer crash when there is a filter and a different root item than the scene, so I don't mind spending more time on this PR to ship all of it together.

cpinter commented on Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyTreeView.cxx:
```diff
@@ -1298,6 +1301,63 @@ bool qMRMLSubjectHierarchyTreeView::multiSelection()
 }
 
 //-----------------------------------------------------------------------------
+vtkIdType qMRMLSubjectHierarchyTreeView::firstSelectedInHierarchyFrom(vtkIdType itemID)
```
I think the method name is not descriptive enough, it doesn't specify first of what is looked for. Also, the convention in SH is that we call "sub-hierarchies" branches. I'd suggest calling this firstSelectedSubjectHierarchyItemInBranch

agirault replied:
great name :+1: I was not happy about the current one.

cpinter commented on Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyTreeView.cxx:
```diff
@@ -1298,6 +1301,63 @@ bool qMRMLSubjectHierarchyTreeView::multiSelection()
 //-----------------------------------------------------------------------------
+vtkIdType qMRMLSubjectHierarchyTreeView::firstSelectedInHierarchyFrom(vtkIdType itemID)
+{
+  // Check self
```
Again, not descriptive enough. I'd say
Check if item itself is selected

cpinter commented on Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyTreeView.cxx:
```diff
@@ -1298,6 +1301,63 @@ bool qMRMLSubjectHierarchyTreeView::multiSelection()
+    return itemID;
+    }
+
+  // Check children
```
Let's make the code more readble. If you add a comment, make it actually explain something
"Look for item in children, recursively"

cpinter commented:
> We're going to need that fixed for our work

That's fine, I just say that this PR contains one well-defined set of features, and considering that the root item issue will probably be hard to fix, I'd make it a logically separate piece of work. Please let's get this PR done first, then we can move on to that bug.

agirault commented:
@cpinter would not showing a popup when the number of items in the treeview is nul be an acceptable workaround?

cpinter commented:
I'm afraid no. If nothing shows up, the user will think the combobox doesn't work. It needs to show something, and not crash. I'm investigating the crash right now, at the least want to see what makes it crash. It's also not very nice that it shows the scene. I think the best would be to show a popup saying "No items matching the filter" if there is a filter and that's why it's empty. It should be easy to add to the popup in qMRMLSubjectHierarchyComboBox::showPopup, but it still won't fix the crash in the tree view I think. Can we pleease add a bug for that and worry about it once the PR is integrated?

cpinter commented:
(Can be a minute after, I'd just like to see it separated. Doing everything in one go will make it harder to scan the history later and seeing what has what effect)

cpinter commented:
The crash comes from here
https://github.com/agirault/Slicer/blob/improve-subjecthierarchy-combobox/Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSortFilterSubjectHierarchyProxyModel.cxx#L421

agirault commented:
Since the crash comes from my changes (flags) I should probably fix it in this pr then? And keep the issue you are talking about (hiding Scene) for a next pr ?

I rebased the commits accordingly so they should not need to be squashed.

Two other issues I get:

1. The popup view is wider than the combobox (see below, on the right side). It seems to be due to [these two lines](
https://github.com/Slicer/Slicer/blob/master/Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx#L288-L289). Any reason they are here? Could I remove them?

<img width="200" alt="screen shot 2017-10-18 at 11 04 16 am" src="https://user-images.githubusercontent.com/4910855/31732009-f1e5c39a-b404-11e7-9504-46a9b7a7a88d.png">

<img width="263" alt="screen shot 2017-10-18 at 1 01 28 pm" src="https://user-images.githubusercontent.com/4910855/31732153-723af510-b405-11e7-91a1-64cb38d0c2a6.png">

2. The combobox does not seem to stretch all the way horizontally (this is in a form layout):

<img width="498" alt="screen shot 2017-10-18 at 10 59 57 am" src="https://user-images.githubusercontent.com/4910855/31732171-8c9adb96-b405-11e7-8980-f35404ba32d9.png">

If I replace any of them by a MRMLNode combobox instead, it will stretch properly without changing anything else:

<img width="560" alt="screen shot 2017-10-18 at 1 19 26 pm" src="https://user-images.githubusercontent.com/4910855/31732654-3da7dfbe-b407-11e7-8ed9-cff5cf1ee63f.png">

So it seems to be only for the subject hierarchy combo box

cpinter commented:
> Since the crash comes from my changes (flags) I should probably fix it in this pr then? And keep the issue you are talking about (hiding Scene) for a next pr ?

Yes, please at least add a check if item is not NULL. That is usually a good idea anyway.

> 1. The popup view is wider than the combobox (see below, on the right side). It seems to be due to these two lines. Any reason they are here? Could I remove them?

They seem to be margins, so if you remove them then at least on some platforms some of the tree view content will be clipped. Please keep them as is for now. This is another issue that can be niely addressed differently. Let's not do everything in one PR! It would be like those "salad" laws that contain unrelated stuff just so that a few wobbly ones can pass as well.

> 2. The combobox does not seem to stretch all the way horizontally

Would it be possible to keep these separate issues in mind and just do whatever was the purpose of this PR? I promise they will be integrated as well :)

agirault commented:
>  It would be like those "salad" laws that contain unrelated stuff just so that a few wobbly ones can pass as well.

Well, if we keep all the commits self-contained and separated like I have, I'm not sure why it's an issue to put all those changes regarding the subject hierarchy combobox in the same pr, but I'll do that.

> Yes, please at least add a check if item is not NULL. That is usually a good idea anyway.

Done, ready for final review/merge then :+1:

cpinter commented:
> Well, if we keep all the commits self-contained and separated like I have, I'm not sure why it's an issue to put all those changes regarding the subject hierarchy combobox in the same pr, but I'll do that.

Existing bugs should be recorded and addressed in Mantis, that's why we have an issue tracker.

That said, I don't think that it makes sense to keep all your commits separate, as some of them didn't even yield a functional revision, and were just later reversed and changed. Please squash at least those commits together that are incremental changes for the same thing (for example fixing the crash: it doesn't make sense to be a separate commit if the crash was caused by another commit within the same PR).

agirault commented:
> for example fixing the crash: it doesn't make sense to be a separate commit if the crash was caused by another commit within the same PR

Yes, I did that when I fixed it: I did an interactive rebase and put it in the `Disable selections of parent items with filters` commit (c8b9ca0)

Let me know if I need to squash more commits together.

cpinter commented:
OK thanks!

agirault commented:
@cpinter @jcfr looks like the doc was properly generated but not published. Any idea why? Is it from my pr? If not can this be merged?

jcfr commented:
>  Any idea why? 

yes - too much data pushed to https://github.com/Slicer/preview.apidocs.slicer.org/tree/gh-pages

I will setup an automatic reset (every Sunday)

> Is it from my pr? 

no 

> If not can this be merged?

Yes - Development of Slicer 4.9.0 is now open.

agirault commented:
@cpinter are you happy with the current state of the pr? Do you want me to merge or should I?

cpinter commented on Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx:
```diff
@@ -64,6 +65,7 @@ class qMRMLSubjectHierarchyComboBoxPrivate
 qMRMLSubjectHierarchyComboBoxPrivate::qMRMLSubjectHierarchyComboBoxPrivate(qMRMLSubjectHierarchyComboBox& object)
   : q_ptr(&object)
   , MaximumNumberOfShownItems(20)
+  , AlignPopupVertically(true)
```
Does this mean that the default is now changed?

agirault replied:
No, the default would align it vertically (selected item overlayed on top of combobox)

cpinter replied:
OK, let's see which one seems better in the long run. Maybe the way you implemented it is more desirable to the users. I don't really have a preference right now, just wanted to know.

cpinter commented on Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyComboBox.cxx:
```diff
@@ -155,6 +159,13 @@ void qMRMLSubjectHierarchyComboBox::setMRMLScene(vtkMRMLScene* scene)
 }
 
 //------------------------------------------------------------------------------
+void qMRMLSubjectHierarchyComboBox::clearCurrentItem()
```
Please call this function also clearSelection. I don't see a reason for this to have a name different from the one in the tree view when we only call that.

agirault replied:
Because treeviews, models etc. deal with **selected** items/indexes : http://doc.qt.io/qt-4.8/qtreeview.html#selectedIndexes

While a combobox only deals with **current** item/index, not "selections" : 
http://doc.qt.io/qt-4.8/qcombobox.html#currentIndex-prop

Thoughts?

cpinter replied:
This function right now is the most basic wrapper imaginable for TreeView::clearSelection, as the only thing it does is calls that. Having a different name for such a wrapper is I think misleading.

Also I think the name clearCurrentItem is not straightforward, as the first thing that occurs to me is that this function clears the _content_ of the current item.

I really think clearSelection is better in every way.

agirault replied:
sounds good, changing that.

cpinter commented on Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyTreeView.cxx:
```diff
@@ -415,7 +418,7 @@ void qMRMLSubjectHierarchyTreeView::setCurrentItems(QList<vtkIdType> items)
     }
 
-  this->selectionModel()->clearSelection();
+  this->clearSelection();
```
Can you please add a comment why calling this is necessary?

agirault replied:
I do not know what motivates the clear before the change of items, this was here prior to my changes, I just replaced the `model::clearSelection()` by the wrapped version of that function in that class.

cpinter replied:
Oh, OK then. Thanks for the explanation! I just thought based on the GitHub diff as I saw it that this is a newly added line, and in that case there must have been a reason for adding it. As it's not a new line, I don't expect you to figure out the reason.

cpinter commented on Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyTreeView.cxx:
```diff
@@ -443,7 +446,7 @@ void qMRMLSubjectHierarchyTreeView::setCurrentItems(vtkIdList* items)
     }
 
-  this->selectionModel()->clearSelection();
+  this->clearSelection();
```
Can you please add a comment why calling this is necessary?

agirault replied:
I do not know what motivates the clear before the change of items, this was here prior to my changes, I just replaced the `model::clearSelection()` by the wrapped version of that function in that class.

cpinter replied:
[See above] Thanks!

cpinter commented on Modules/Loadable/SubjectHierarchy/Widgets/qMRMLSubjectHierarchyTreeView.h:
```diff
@@ -176,6 +177,9 @@ protected slots:
   /// (e.g. after tree was updated in the model from the subject hierarchy)
   virtual void updateRootItem();
 
+  /// Called when a subject hierarchy item is modified
```
This is evident. Please explain what it does with the modified item.

cpinter commented:
I added some minor comments. After addressing those you can integrate if you have the privileges to do that.

Then we can work on the bugs you caught. It's good to see what change has what effect, even better if there is a nightly between the PR and the bugfix so that we can use those later if necessary.

Thanks again!

agirault commented:
Thank you for the reviews.

Integrated in:
- [r26495](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=26495)
- [r26496](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=26496)
- [r26497](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=26497)
- [r26498](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=26498)
- [r26499](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=26499)

cpinter commented:
Great! Thanks for working on this and dealing with my nitpicks!
